src/snarky-typescript/src/index.ts

```diff
@@ -0,0 +1,517 @@
+  read(x: F): F;
+  read(x: Var): F;
+
+  new (): Var;
+  new (options: {isPublic: boolean}): Var;
+
+  new (x: FieldConvertible): Var;
+  new (x: F): Var;
+  new (x: Var): Var;
+  new (f: () => FieldConvertible | F | Var): Var;
+
+  new (x: FieldConvertible, options: {isPublic: boolean}): Var;
+  new (x: F, options: {isPublic: boolean}): Var;
+  new (x: Var, options: {isPublic: boolean}): Var;
+  new (f: () => F, options: {isPublic: boolean}): Var;
+  new (f: () => FieldConvertible, options: {isPublic: boolean}): Var;
+
+  new (x: FieldConvertible, isPublic: boolean): Var;
+  new (x: F, isPublic: boolean): Var;
+  new (x: Var, isPublic: boolean): Var;
+  new (f: () => F, isPublic: boolean): Var;
+  new (f: () => FieldConvertible, isPublic: boolean): Var;
+}
+
+/* TODO: Fill this in with the default implementation */
+abstract class BoolVar<F, Circuit, V extends CircuitVar<F, Circuit>> {
+  abstract not(this: this): this;
+  abstract and(this: this, y: boolean | this): this;
+  abstract or(this: this, y: boolean | this): this;
+
+  abstract equals(this: this, y: boolean | this): this;
+  abstract assertEquals(this: this, y: boolean | this): void;
+  abstract assertTrue(this: this): void;
+  abstract assertFalse(this: this): void;
+
+  abstract if_<Var extends AsFieldElements<V, Var>>(
+    this: this,
+    then_: Var,
+    else_: Var
+  ): Var;
+  abstract if_<Var extends AsFieldElements<V, Var>>(
+    this: this,
+    args: {
+      then_: Var;
+      else_: Var;
+    }
+  ): Var;
+
+  abstract toField(this: this): V;
+  abstract toFieldElements(this: this): [V];
+  abstract ofFieldElements(x: V[]): this;
+
+  abstract read(this: this): boolean;
+}
+
+interface BoolVarCtor<
+  F,
+  Circuit,
+  V extends CircuitVar<F, Circuit>,
+  BVar extends BoolVar<F, Circuit, V>
+> {
+  not(x: BVar | boolean): BVar;
+  and(x: BVar | boolean, y: BVar | boolean): BVar;
+  or(x: BVar | boolean, y: BVar | boolean): BVar;
+
+  countTrue(xs: (BVar | boolean)[]): V;
+  countFalse(xs: (BVar | boolean)[]): V;
+
+  allTrue(xs: (BVar | boolean)[]): BVar;
+  allFalse(xs: (BVar | boolean)[]): BVar;
+  anyTrue(xs: (BVar | boolean)[]): BVar;
+  anyFalse(xs: (BVar | boolean)[]): BVar;
+
+  equals(x: BVar | boolean, y: BVar | boolean): BVar;
+
+  assertTrue(xs: BVar | boolean): BVar;
+  assertFalse(xs: BVar | boolean): BVar;
+  assertAllTrue(xs: (BVar | boolean)[]): BVar;
+  assertAllFalse(xs: (BVar | boolean)[]): BVar;
+  assertAnyTrue(xs: (BVar | boolean)[]): BVar;
+  assertAnyFalse(xs: (BVar | boolean)[]): BVar;
+  assertEquals(x: BVar | boolean, y: BVar | boolean): void;
+
+  toField(x: BVar | boolean): V;
+
+  ofField(x: FieldConvertible): BVar;
+  ofField(x: F): BVar;
+  ofField(x: V): BVar;
+
+  toFieldElements(x: BVar | boolean): [V];
+
+  ofFieldElements(x: (FieldConvertible | F | V)[]): BVar;
+
+  read(x: BVar | boolean): boolean;
+
+  Unsafe: {
+    ofField(x: FieldConvertible): BVar;
+    ofField(x: F): BVar;
+    ofField(x: V): BVar;
```

Comment:
Why is `ofField` unsafe, but `ofFieldElements` is not?

Reply:
I couldn't think of a better way to make this work: basically `ofFieldElements` is the equivalent of half of `Typ.t` in OCaml snarky (to implement `AsFieldElements`), and I didn't want users to have to write `Bool.Unsafe` to instantiate a boolean variables; OTOH, it *is* unsafe to uplift a field element to a boolean without checking first, so I provided this API to indicate that you can do it, but only if you're careful.

Perhaps there's a nice way to mark the to/fromFieldElements as unsafe?